src/wallet/verify_wallet_flow.c

```diff
@@ -0,0 +1,253 @@
+extern char *NUMBERS;
+
+/*****************************************************************************
+ * PRIVATE MACROS AND DEFINES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * PRIVATE TYPEDEFS
+ *****************************************************************************/
+typedef enum {
+  PIN_INPUT,
+  TAP_CARD_FLOW,
+  VERIFY_SHARES,
+  COMPLETED,
+  COMPLETED_WITH_ERRORS,
+  TIMED_OUT,
+  EARLY_EXIT,
+  EXIT,
+} verify_wallet_state_e;
+
+/*****************************************************************************
+ * STATIC FUNCTION PROTOTYPES
+ *****************************************************************************/
+/**
+ * @brief State handler for the verify wallet flow on the X1 vault
+ *
+ * @param current_state The current state of the flow which needs to be executed
+ * @return verify_wallet_state_e The next state of the flow, based on processing
+ * the current state handler
+ */
+verify_wallet_state_e verify_wallet_state_handler(
+    verify_wallet_state_e current_state);
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+verify_wallet_state_e verify_wallet_state_handler(
+    verify_wallet_state_e current_state) {
+  verify_wallet_state_e next_state = EXIT;
+
+  switch (current_state) {
+    case PIN_INPUT: {
+      if (!WALLET_IS_PIN_SET(wallet.wallet_info)) {
+        next_state = TAP_CARD_FLOW;
+        break;
+      }
+
+      WALLET_SET_PIN(wallet_for_flash.wallet_info);
+      WALLET_SET_PIN(wallet.wallet_info);
+
+      input_text_init(ALPHA_NUMERIC, ui_text_enter_pin, 4, DATA_TYPE_PIN, 8);
+      next_state = get_state_on_input_scr(PIN_INPUT, EARLY_EXIT, TIMED_OUT);
+
+      if (PIN_INPUT != next_state) {
+        break;
+      }
+
+      sha256_Raw((uint8_t *)flow_level.screen_input.input_text,
+                 strnlen(flow_level.screen_input.input_text,
+                         sizeof(flow_level.screen_input.input_text)),
+                 wallet_credential_data.password_single_hash);
+      sha256_Raw(wallet_credential_data.password_single_hash,
+                 SHA256_DIGEST_LENGTH,
+                 wallet.password_double_hash);
+      memzero(flow_level.screen_input.input_text,
+              sizeof(flow_level.screen_input.input_text));
+      next_state = TAP_CARD_FLOW;
+      break;
+    }
+
+    case TAP_CARD_FLOW: {
+      if (true == card_flow_verify_wallet()) {
```

Comment:
wrong PIN input can happen here. add the reasoning for that case; how we want to handle it.

Reply:
updated

Reply:
Only added error messages, not added retry attempt